### PR TITLE
Add gc/8.0.4 recipe

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **gc/8.0.4**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/gc/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        if tools.Version(self.version) <= "8.0.4":
+            del self.options.throw_bad_alloc_library
```
Comment (uilianries) [inline recipes/gc/all/conanfile.py]:
```suggestion
            del self.options.throw_bad_alloc_library
        if not self.options.cplusplus:
            del self.settings.compiler.libcxx
            del self.settings.compiler.cppstd
```

Both libgc and libcord don't require C++ lib

File: recipes/gc/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        assert index > 0
+        return readme_md[index+len(copyright_header):]
+
+    def package(self):
+        tools.save(os.path.join(self.package_folder, "licenses", "COPYRIGHT"), self._extract_copyright())
+        cmake = self._configure_cmake()
+        cmake.install()
+
+        # tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        # tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
```
Comment (uilianries) [inline recipes/gc/all/conanfile.py]:
```suggestion
```

File: recipes/gc/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        libs.append("gc")
+        return libs
+
+    def package_info(self):
+        self.cpp_info.names["pkg_config"] = "bdw-gc"
+        self.cpp_info.libs = self._libs
+        self.cpp_info.defines = ["GC_DLL" if self.options.shared else "GC_NOT_DLL"]
+        if not self.options.shared:
+            if self.settings.os == "Linux":
+                self.cpp_info.system_libs = ["pthread"]
```
Comment (uilianries) [inline recipes/gc/all/conanfile.py]:
```suggestion
                self.cpp_info.system_libs = ["pthread", "dl"]
```

libgc and libcord require libdl

File: recipes/gc/all/patches/update-cmake.patch [hunk 4]
```diff
@@ -0,0 +1,834 @@
+--- CMakeLists.txt
```
Comment (uilianries) [inline recipes/gc/all/patches/update-cmake.patch]: This is a complete new CMake file. Did you implement it, or did you find it in some repository?

File: recipes/gc/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            tools.patch(**patch)
+
+    def build(self):
+        self._patch_sources()
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def _extract_copyright(self):
+        readme_md = open(os.path.join(self._source_subfolder, "README.md")).read()
+        copyright_header = "## Copyright & Warranty\n"
```
Comment (uilianries) [inline recipes/gc/all/conanfile.py]: Same content than https://www.hboehm.info/gc/license.txt. Okay.

File: recipes/gc/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        index = readme_md.find(copyright_header)
+        assert index > 0
```
Comment (uilianries) [inline recipes/gc/all/conanfile.py]:
```suggestion
        if index == 0:
            raise ConanException("Could not extract license from README file.")
```

You need to import ConanException too.

File: recipes/gc/all/patches/update-cmake.patch (hunk 4 above)
Comment (madebr) [reply recipes/gc/all/patches/update-cmake.patch]: This is the cmake script form current master.
I could not let autotools create a shared Visual Studio build.

Comment (uilianries) [reply recipes/gc/all/patches/update-cmake.patch]: I see, so it's totally valid. thanks.

File: recipes/gc/all/conanfile.py (hunk 6 above)
Comment (madebr) [reply recipes/gc/all/conanfile.py]: `find` returns `-1` when a string cannot be found.

File: recipes/gc/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
```
Comment (madebr) [inline recipes/gc/all/conanfile.py]:
```suggestion
        if index == -1:
            raise ConanException("Could not extract license from README file.")
```

File: recipes/gc/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
+from conans import CMake, ConanFile, tools
```
Comment (madebr) [inline recipes/gc/all/conanfile.py]:
```suggestion
from conans import CMake, ConanFile, tools
from conans import ConanException
```

File: recipes/gc/all/patches/update-cmake.patch (hunk 4 above)
Comment (madebr) [reply recipes/gc/all/patches/update-cmake.patch]: Thanks for reviewing btw.

File: recipes/gc/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+from conans import ConanException
```
Comment (madebr) [inline recipes/gc/all/conanfile.py]:
```suggestion
from conans.errors import ConanException
```

File: recipes/gc/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
+from conans import CMake, ConanFile, tools
+from conans.errors import ConanException
+import os
+
+
+class GcConan(ConanFile):
+    name = "gc"
```
Comment (uilianries) [inline recipes/gc/all/conanfile.py]:
```suggestion
    name = "bdwgc"
```

https://github.com/microsoft/vcpkg/tree/master/ports/bdwgc

Comment (madebr) [reply recipes/gc/all/conanfile.py]: Done

File: recipes/bdwgc/all/patches/update-cmake.patch [hunk 11]
```diff
@@ -0,0 +1,834 @@
+--- CMakeLists.txt
```
Comment (SSE4) [inline recipes/bdwgc/all/patches/update-cmake.patch:1]: many changes, did you take this patch somewhere or created by hand?
what's the summary of the changes?
consider renaming to something more specific that "update cmake" or adding comment(s)

Comment (madebr) [reply recipes/bdwgc/all/patches/update-cmake.patch:1]: This is the current master's cmake script.

**Pull request merged**